src/drivers/imx/sai.c

```diff
@@ -74,32 +110,39 @@ static void sai_stop(struct dai *dai, int direction)
 
 	uint32_t xcsr = 0U;
 
+	/* Disable DMA request */
 	dai_update_bits(dai, REG_SAI_XCSR(direction),
 			REG_SAI_CSR_FRDE, 0);
+
+	/* Transmit/Receive data channel disable */
+	dai_update_bits(dai, REG_SAI_XCR3(direction),
+			REG_SAI_CR3_TRCE_MASK,
+			REG_SAI_CR3_TRCE(0));
+
+	/* Disable interrupts */
 	dai_update_bits(dai, REG_SAI_XCSR(direction),
 			REG_SAI_CSR_XIE_MASK, 0);
 
-	/* Check if the opposite direction is also disabled */
-	xcsr = dai_read(dai, REG_SAI_XCSR(!direction));
-	if (!(xcsr & REG_SAI_CSR_FRDE)) {
-		/* Disable both directions and reset their FIFOs */
-		dai_update_bits(dai, REG_SAI_TCSR, REG_SAI_CSR_TERE, 0);
-		poll_for_register_delay(dai_base(dai) + REG_SAI_TCSR,
-					REG_SAI_CSR_TERE, 0, 100);
-
-		dai_update_bits(dai, REG_SAI_RCSR, REG_SAI_CSR_TERE, 0);
-		poll_for_register_delay(dai_base(dai) + REG_SAI_RCSR,
+	/* Disable transmitter/receiver */
+	if (direction == DAI_DIR_CAPTURE) {
+		dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_CAPTURE), REG_SAI_CSR_TERE, 0);
+		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_CAPTURE),
 					REG_SAI_CSR_TERE, 0, 100);
-
-		/* Software Reset for both Tx and Rx */
-		dai_update_bits(dai, REG_SAI_TCSR, REG_SAI_CSR_SR,
-				REG_SAI_CSR_SR);
-		dai_update_bits(dai, REG_SAI_RCSR, REG_SAI_CSR_SR,
-				REG_SAI_CSR_SR);
-
-		/* Clear SR bit to finish the reset */
-		dai_update_bits(dai, REG_SAI_TCSR, REG_SAI_CSR_SR, 0U);
-		dai_update_bits(dai, REG_SAI_RCSR, REG_SAI_CSR_SR, 0U);
+		/* Check if the opposite direction is also disabled */
+		xcsr = dai_read(dai, REG_SAI_XCSR(DAI_DIR_PLAYBACK));
+		if (!(xcsr & REG_SAI_CSR_FRDE)) {
+			dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_PLAYBACK), REG_SAI_CSR_TERE, 0);
+			poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_PLAYBACK),
+						REG_SAI_CSR_TERE, 0, 100);
+		}
+	} else {
+		/* Check if the opposite direction is also disabled */
+		xcsr = dai_read(dai, REG_SAI_XCSR(DAI_DIR_CAPTURE));
+		if (!(xcsr & REG_SAI_CSR_FRDE)) {
+			dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_PLAYBACK), REG_SAI_CSR_TERE, 0);
+			poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_PLAYBACK),
+						REG_SAI_CSR_TERE, 0, 100);
+		}
 	}
```

Comment:
At least this part can be rewritten like this:

```
	/* Disable transmitter/receiver */
	if (direction == DAI_DIR_CAPTURE) {
		dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_CAPTURE), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_CAPTURE),
					REG_SAI_CSR_TERE, 0, 100);
	}
	
	
	/* Check if the opposite direction is also disabled */
	xcsr = dai_read(dai, REG_SAI_XCSR(!direction));
	if (!(xcsr & REG_SAI_CSR_FRDE)) {
		dai_update_bits(dai, REG_SAI_XCSR(!direction), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(!direction),
						REG_SAI_CSR_TERE, 0, 100);
	}
```

Reply:
It can be written like that only if this
```
if (!(xcsr & REG_SAI_CSR_FRDE)) {
		dai_update_bits(dai, REG_SAI_XCSR(!direction), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(!direction),
						     REG_SAI_CSR_TERE, 0, 100);
}
```
changes to this:

```
if (!(xcsr & REG_SAI_CSR_FRDE)) {
		dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_PLAYBACK), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_PLAYBACK),
						     REG_SAI_CSR_TERE, 0, 100);
}
```

Reply:
As per our discussion, I've kept the original code for a clear/better understanding.